### PR TITLE
[rhidp] decommission app interface flavor

### PR DESCRIPTION
**Pull request by chassing**
Refactor the RHIDP capability (`rhidp.sso-client` & `rhidp.ocm-oidc-idp`) and remove the deprecated App-Interface flavor. The capability now only supports clusters labeled with `sre-capabilities.rhidp.*` labels.

The capability now supports additional OCM labels to customize the behavior:

* The single `sre-capabilities.rhidp` label is now deprecated but still supported
* `sre-capabilities.rhidp.status` replaces the `sre-capabilities.rhidp` label and has these available options:
  * `enabled`: Enable RedHat SSO authentication for the cluster
  * `disabled`: Disable previously configured RedHat SSO authentication
  * `enforced`: Like `enabled` but also remove all other manually configured IDPs
  * `sso-client-only` - Create an SSO client, but don't manage the OIDC cluster config
* `sre-capabilities.rhidp.name`: Configure the authentication name on the console login page; default is `redhat-sso`. Spaces will be replaced with dashes.
* `sre-capabilities.rhidp.issuer`: Configure the OIDC issuer URL

Hints for the reviewer:

* 8f86ec66964aa: This PR removes the `ClusterV1` dependency; therefore, the capability can't use `OCMMap` anymore, therefore the need to introduce `utils.ocm.identity_providers`
*  2nd commit refactors the capability
*  3rd commit refactors the tests

Ticket: [APPSRE-8126](https://issues.redhat.com/browse/APPSRE-8126)
Schemas: https://github.com/app-sre/qontract-schemas/pull/510

### CODE REVIEW

File: reconcile/utils/ocm/base.py [hunk 1]
```diff
@@ -16,12 +18,15 @@
 CAPABILITY_MANAGE_CLUSTER_ADMIN = "capability.cluster.manage_cluster_admin"
 
 
-class OCMModelLink(BaseModel):
+class OCMModelLinkSimple(BaseModel):
```
Comment (geoberle) [inline reconcile/utils/ocm/base.py:21]: suggestion: `OCMCollectionLink`
if the idea is missing, this kind of reference is for a collection of models.

File: reconcile/rhidp/ocm_oidc_idp/base.py [hunk 2]
```diff
@@ -21,48 +16,59 @@
...
-        ocm_map = OCMMap(
-            clusters=[cluster.dict(by_alias=True) for cluster in clusters],
-            integration=integration_name,
-            settings=settings,
-        )
-
         try:
             # run
-            current_state = fetch_current_state(ocm_map, clusters)
+            current_state = fetch_current_state(ocm_api, clusters)
```
Comment (geoberle) [inline reconcile/rhidp/ocm_oidc_idp/base.py:65]: ❤️

**Review by geoberle — APPROVED** [review]
LGTM

File: reconcile/rhidp/ocm_oidc_idp/base.py [hunk 3]
```diff
@@ -21,48 +16,59 @@
...
     integration_name: str,
     ocm_environment: str,
-    clusters: Iterable[ClusterV1],
+    clusters: Iterable[Cluster],
     secret_reader: SecretReaderBase,
+    ocm_api: OCMBaseClient,
     vault_input_path: str,
     dry_run: bool,
-    managed_idps: Optional[list[str]] = None,
+    managed_idps: list[str] | None = None,
```
Comment (hemslo) [inline reconcile/rhidp/ocm_oidc_idp/base.py:55]: Is this the new preferred typing for optional field?

File: reconcile/rhidp/ocm_oidc_idp/base.py [hunk 4]
```diff
@@ -82,138 +88,134 @@ def run(
...
-            logging.debug(f"Skipping removal of unmanged '{idp.name}' IDP.")
-            continue
-        logging.info(["remove_oidc_idp", idp.cluster, idp.name])
-        if not idp.id:
-            logging.error(
-                "No identity provider id was given. This should never ever happen!"
+    diff_result = diff_iterables(
+        current_state,
+        desired_state,
+        key=lambda idp_state: hash(
```
Comment (hemslo) [inline reconcile/rhidp/ocm_oidc_idp/base.py:172]: tuple should be enough here, like

```python
        key=lambda idp_state: (
            idp_state.cluster.organization_id,
            idp_state.cluster.name,
            idp_state.idp.type,
            idp_state.idp.name,
        ),

```

File: reconcile/utils/ocm/identity_providers.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        for idp_dict in ocm_api.get_paginated(
+            api_path=ocm_cluster.identity_providers.href
+        ):
+            match idp_dict["type"]:
+                case "OpenIDIdentityProvider":
+                    yield OCMOIdentityProviderOidc(**idp_dict)
+                case "GithubIdentityProvider":
+                    yield OCMOIdentityProviderGithub(**idp_dict)
+                case _:
+                    yield OCMOIdentityProvider(**idp_dict)
```
Comment (hemslo) [inline reconcile/utils/ocm/identity_providers.py:30]: Nice method extraction using pattern match and `yield`! ❤️

**Review by hemslo — APPROVED** [review]
LGTM!

File: reconcile/rhidp/ocm_oidc_idp/base.py (hunk 3 above)
Comment (chassing) [reply reconcile/rhidp/ocm_oidc_idp/base.py:55]: This came with Python 3.11, and I like it more :)

File: reconcile/utils/ocm/base.py (hunk 1 above)
Comment (chassing) [reply reconcile/utils/ocm/base.py]: Applied in [f1a1942](https://github.com/app-sre/qontract-reconcile/pull/3775/commits/f1a1942cbf1e3a0a3b438ea3613e9e1187b62d37)

File: reconcile/rhidp/ocm_oidc_idp/base.py (hunk 4 above)
Comment (chassing) [reply reconcile/rhidp/ocm_oidc_idp/base.py]: Applied in [f1a1942](https://github.com/app-sre/qontract-reconcile/pull/3775/commits/f1a1942cbf1e3a0a3b438ea3613e9e1187b62d37)

**Pull request merged**